post_request_task/task.py

```diff
@@ -29,48 +31,52 @@ def _start_queuing_tasks(**kwargs):
     If not called, tasks are delayed normally (so tasks still function without
     having to call _send_tasks_and_stop_queuing() manually when we're outside
     the request-response cycle.."""
-    _locals.__dict__['queue_tasks'] = True
+    _locals.queue_tasks = True
 
 
 def _stop_queuing_tasks(**kwargs):
     """Stops queuing tasks for this thread.
 
     Not supposed to be called directly, only useful for tests cleanup."""
-    _locals.__dict__.pop('queue_tasks', None)
+    _locals.queue_tasks = False
 
 
 def is_task_queuing_enabled_for_this_thread():
     """Returns whether post request task queuing is enabled for this thread."""
-    return _locals.__dict__.get('queue_tasks', False)
+    return _locals.queue_tasks
 
 
 def _send_tasks_and_stop_queuing(**kwargs):
     """Sends all delayed Celery tasks and stop queuing new ones for now."""
-    _locals.__dict__.pop('queue_tasks', None)
+    log.info('Stopping queueing tasks and sending already queued ones.')
+    _stop_queuing_tasks()
     queue = _get_task_queue()
     while queue:
-        cls, args, kwargs = queue.pop(0)
-        cls.original_apply_async(*args, **kwargs)
+        task, args, kwargs, extrakw = queue.pop(0)
+        task.original_apply_async(args=args, kwargs=kwargs, **extrakw)
 
 
 def _discard_tasks(**kwargs):
     """Discards all delayed Celery tasks."""
-    _get_task_queue()[:] = []
+    log.info('Discarding %d queued tasks.', len(_locals.task_queue))
+    _locals.task_queue = []
 
 
 def _append_task(t):
     """Append a task to the queue.
 
-    Expected argument is a tuple of the (task class, args, kwargs).
+    Expected argument is a tuple of the following form:
+    (task class, args, kwargs, extra kwargs).
 
     This doesn't append to queue if the argument is already in the queue.
 
     """
-    queue = _get_task_queue()
+    queue = _locals.task_queue
     if t not in queue:
+        log.debug('Appended new task to the queue: %s.', t)
         queue.append(t)
     else:
-        log.debug('Removed duplicate task: %s' % (t,))
+        log.debug('Did not append new duplicate task to the queue: %s.', t)
```

Comment:
If it's new can it also be a duplicate? This reads strangely to me. Dropping "new" makes more sense I think.